ops/internal/config/chain.go

```diff
@@ -182,34 +183,93 @@ type AddressesWithRoles struct {
 	Roles
 }
 
+func CreateAddressesWithRolesFromFetcher(addresses script.Addresses, roles script.Roles) AddressesWithRoles {
+	addressesWithRoles := AddressesWithRoles{
+		Addresses: Addresses{
+			AddressManager:                    NewChecksummedAddress(addresses.AddressManager),
+			L1CrossDomainMessengerProxy:       NewChecksummedAddress(addresses.L1CrossDomainMessengerProxy),
+			L1ERC721BridgeProxy:               NewChecksummedAddress(addresses.L1ERC721BridgeProxy),
+			L1StandardBridgeProxy:             NewChecksummedAddress(addresses.L1StandardBridgeProxy),
+			L2OutputOracleProxy:               NewChecksummedAddress(addresses.L2OutputOracleProxy),
+			OptimismMintableERC20FactoryProxy: NewChecksummedAddress(addresses.OptimismMintableERC20FactoryProxy),
+			OptimismPortalProxy:               NewChecksummedAddress(addresses.OptimismPortalProxy),
+			SystemConfigProxy:                 NewChecksummedAddress(addresses.SystemConfigProxy),
+			ProxyAdmin:                        NewChecksummedAddress(addresses.OpChainProxyAdmin),
+			SuperchainConfig:                  NewChecksummedAddress(addresses.SuperchainConfig),
+			AnchorStateRegistryProxy:          NewChecksummedAddress(addresses.AnchorStateRegistryProxy),
+			DisputeGameFactoryProxy:           NewChecksummedAddress(addresses.DisputeGameFactoryProxy),
+			FaultDisputeGame:                  NewChecksummedAddress(addresses.FaultDisputeGame),
+			MIPS:                              NewChecksummedAddress(addresses.Mips),
+			PermissionedDisputeGame:           NewChecksummedAddress(addresses.PermissionedDisputeGame),
+			PreimageOracle:                    NewChecksummedAddress(addresses.PreimageOracle),
+		},
+		Roles: Roles{
+			SystemConfigOwner: NewChecksummedAddress(roles.SystemConfigOwner),
+			ProxyAdminOwner:   NewChecksummedAddress(roles.OpChainProxyAdminOwner),
+			Guardian:          NewChecksummedAddress(roles.Guardian),
+			Challenger:        NewChecksummedAddress(roles.Challenger),
+			Proposer:          NewChecksummedAddress(roles.Proposer),
+			UnsafeBlockSigner: NewChecksummedAddress(roles.UnsafeBlockSigner),
+			BatchSubmitter:    NewChecksummedAddress(roles.BatchSubmitter),
+		},
+	}
+	// Hack until we separate the permissioned and permissionless WETH proxies
+	if addresses.DelayedWETHPermissionlessGameProxy != (common.Address{}) {
+		addressesWithRoles.Addresses.DelayedWETHProxy = NewChecksummedAddress(addresses.DelayedWETHPermissionlessGameProxy)
+	} else {
+		addressesWithRoles.Addresses.DelayedWETHProxy = NewChecksummedAddress(addresses.DelayedWETHPermissionedGameProxy)
+	}
+
+	return addressesWithRoles
+}
+
 func (a AddressesWithRoles) MarshalJSON() ([]byte, error) {
-	// Create a map to hold all fields
-	allFields := make(map[string]*ChecksummedAddress)
+	// Create a map to hold non-zero address fields
+	allFields := make(map[string]string)
+
+	// Declare processStruct variable first to allow recursion
+	var processStruct func(interface{})
```

Comment:
What is this doing?

Reply:
Basically what the comment says. The recursion below doesn't work without first declaring the var

Reply:
Sorry, I meant the entire method